mm2src/coins/eth.rs

```diff
@@ -7205,3 +7215,39 @@ impl CommonSwapOpsV2 for EthCoin {
         self.derive_htlc_pubkey_v2(swap_unique_data).to_bytes()
     }
 }
+
+#[cfg(all(feature = "for-tests", not(target_arch = "wasm32")))]
+impl EthCoin {
+    pub async fn set_coin_type(&self, new_coin_type: EthCoinType) -> EthCoin {
+        let coin = EthCoinImpl {
+            ticker: self.ticker.clone(),
+            coin_type: new_coin_type,
+            priv_key_policy: self.priv_key_policy.clone(),
+            derivation_method: Arc::clone(&self.derivation_method),
+            sign_message_prefix: self.sign_message_prefix.clone(),
+            swap_contract_address: self.swap_contract_address,
+            swap_v2_contracts: self.swap_v2_contracts,
+            fallback_swap_contract: self.fallback_swap_contract,
+            contract_supports_watchers: self.contract_supports_watchers,
+            web3_instances: AsyncMutex::new(self.web3_instances.lock().await.clone()),
+            decimals: self.decimals,
+            history_sync_state: Mutex::new(self.history_sync_state.lock().unwrap().clone()),
+            required_confirmations: AtomicU64::new(
+                self.required_confirmations.load(std::sync::atomic::Ordering::SeqCst),
+            ),
+            swap_txfee_policy: Mutex::new(self.swap_txfee_policy.lock().unwrap().clone()),
+            max_eth_tx_type: self.max_eth_tx_type,
+            ctx: self.ctx.clone(),
+            chain_id: self.chain_id,
+            trezor_coin: self.trezor_coin.clone(),
+            logs_block_range: self.logs_block_range,
+            address_nonce_locks: Arc::clone(&self.address_nonce_locks),
+            erc20_tokens_infos: Arc::clone(&self.erc20_tokens_infos),
+            nfts_infos: Arc::clone(&self.nfts_infos),
+            platform_fee_estimator_state: Arc::clone(&self.platform_fee_estimator_state),
+            gas_limit: EthGasLimit::default(),
+            abortable_system: self.abortable_system.create_subsystem().unwrap(),
+        };
+        EthCoin(Arc::new(coin))
+    }
```

Comment:
Can't we mutate `self` here like this:

```rs
    pub fn set_coin_type(&mut self, new_coin_type: EthCoinType) { self.coin_type = new_coin_type; }
```

Reply:
ah I remembered, i did the above due to such obstacles:

```
error[E0594]: cannot assign to data in dereference of `eth::EthCoin`
    --> mm2src/coins/eth.rs:7221:67
     |
7221 |     pub fn set_coin_type(&mut self, new_coin_type: EthCoinType) { self.coin_type = new_coin_type; }
     |                                                                   ^^^^^^^^^^^^^^ cannot assign
     |
     = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `eth::EthCoin`

```

```
error[E0277]: the trait bound `eth::EthCoinImpl: std::clone::Clone` is not satisfied
    --> mm2src/coins/eth.rs:7225:43
     |
7225 |             let mut inner = Arc::make_mut(&mut self.0);
     |                             ------------- ^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `eth::EthCoinImpl`
     |                             |
     |                             required by a bound introduced by this call
     |
note: required by a bound in `std::sync::Arc::<T>::make_mut`
```